apps/dashboard/lib/rclone_util.rb

```diff
@@ -0,0 +1,147 @@
+      if s.success?
+        o.lines.any?{ |l| l.strip == path.basename.to_s}
+      else
+        raise StandardError, "Error checking info for path: #{e}"
+      end
+    end
+
+    def mime_type(remote, path)
+      # Check first is it is a directory, to avoid strange lsjson behaviour
+      # when a directory contains a file with the same name as the directory.
+      # `rclone lsjson remote:/name` and `rlcone lsjson remote:/name/name`
+      # returns only the info for remote:/name/name
+      if directory?(remote, path)
+        "inode/directory"
+      else
+        files = ls(remote, path)
+        files.find { |file| file["Path"] == path.basename.to_s }["MimeType"]
+      end
+    end
+
+    def cat(remote, path, &block)
+      full_path = "#{remote}:#{path}"
+      # Read the file in 32kb chunks
+      if block_given?
+        run_popen(rclone_cmd, "cat", full_path) do |o|
+          while data = o.read(32768)
+            yield data
+          end
+        end
+      else
+        # Read the whole file
+        o, e, s = rclone("cat", full_path)
+        if s.success?
+          o
+        else
+          raise StandardError, "Error reading file #{full_path}: #{e}"
+        end
+      end
+    end
+
+    def touch(remote, path)
+      full_path = "#{remote}:#{path}"
+      o, e, s = rclone("touch", full_path)
+      if !s.success?
+        raise StandardError, "Error creating file: #{e}"
+      end
+    end
+
+    def mkdir(remote, path)
+      full_path = "#{remote}:#{path}"
+      o, e, s = rclone("mkdir", full_path)
+      # TODO: detect when Rclone remote doesn't support empty dirs (e.g. S3) and
+      # raise a suitable error message.
+      if !s.success?
+        raise StandardError, "Error creating directory: #{e}"
+      end
+    end
+
+    def write(remote, path, content)
+      full_path = "#{remote}:#{path}"
+      # Write to a file on the remote by passing the file contents in stdin
+      o, e, s = rclone("rcat", full_path, stdin_data: content)
+      if !s.success?
+        raise StandardError, "Error writing file: #{e}"
+      end
+    end
+
+    def moveto(remote, path, src)
+      full_path = "#{remote}:#{path}"
+      # Move file src on the local filesystem to full_path on the remote
+      o, e, s = rclone("moveto", src, full_path)
+      if !s.success?
+        raise StandardError, "Error moving file: #{e}"
+      end
+    end
+
+    def remote_type(remote)
+      # Get the rclone remote type (e.g. s3) for a single remote
+      o, e, s = rclone("listremotes", "--long")
+      if s.success?
+        remote = o.lines.grep(/^#{Regexp.escape(remote)}:/).first
+        return nil if remote.nil?
+
+        type = remote.split(":")[1].strip
+      else
+        raise StandardError, "Error getting information about remote: #{e}"
+      end
+    end
+
+    def rclone_cmd
+      # TODO: Make this configurable
+      "rclone"
+    end
+
+    def rclone(*args)
+      Open3.capture3(rclone_cmd, *args)
```

Comment:
We should probably cast these *args here with `ShellWords.escape`.  I don't know for sure if a malicious command would pass through `validate_path!` but it looks like it could if you didn't have any allowlists (which I'm guessing most sites don't).

The premise being you could send `/some/path; cp ~/.ssh/* /my/malicous/folder`. I'm testing now whether that's possible at all or if I'm just scared.

Reply:
I actually can trigger this blank page error with a URL that's `/<remote>/<directory that for sure doesn't exist>`. Just edit the URL to a folder you're sure doesn't exist.

Reply:
> We should probably cast these *args here with `ShellWords.escape`. I don't know for sure if a malicious command would pass through `validate_path!` but it looks like it could if you didn't have any allowlists (which I'm guessing most sites don't).
> 
> The premise being you could send `/some/path; cp ~/.ssh/* /my/malicous/folder`. I'm testing now whether that's possible at all or if I'm just scared.

As far as I know that is not necessary as passing the arguments in that format spawns the process directly without a shell.
Small test:
```
[rkarlsso@bofink ~]$ irb
3.0.2 :001 > require "open3"
 => true 
3.0.2 :002 > Open3.capture3("cat", "file.txt; echo sometext")
 => 
["",
 "cat: 'file.txt; echo sometext': No such file or directory\n",
 #<Process::Status: pid 38619 exit 1>] 
```

> I actually can trigger this blank page error with a URL that's `/<remote>/<directory that for sure doesn't exist>`. Just edit the URL to a folder you're sure doesn't exist.

Turns out there was a small error in the logic for checking if it is a directory or not, so directories that don't exist defaulted to files, which was handled by show_file. Fixed it so now it properly shows the files app with a message that it doesn't exist.

Reply:
Added some additional parsing of the Rclone exit code there. Most likely something similar could be done in the other functions too, though the [exit codes](https://rclone.org/docs/#exit-code) seems mostly useful for catching directories and files that don't exist in this case, as most other error codes are pretty vague.

Reply:
Added even more error checking. Not sure how much more we can do at the moment. Would be nice to have good messages about authentication errors, but those messages vary a lot between the Rclone remote types.

Reply:
Extra checks are OK, but it seems to me it should have hit that last else block that existed at that time?

> Turns out there was a small error in the logic for checking if it is a directory or not, so directories that don't exist defaulted to files, which was handled by show_file.

Really I'm just worried about crashing like this with the stream writing and all that if/when exceptions happen they're handled.

Reply:
Last else block wasn't hit as we listed directories in the parent, which succeeded (but now we actually verify that the file exists there too). Trying to access something like `/<remote>/<directory>/<directory>` would have raised the error correctly, which would have been caught by the FilesController. Once an error finally was raised (in show_remote_file) due to the file not existing it was too late as the response was already being sent.

Reply:
I'm going to take a quick look at this, but I think we're good to go. We squash + commit PRs so if you want to squash this into 1 commit so you can write the commit message please do so.  Otherwise I'll write the commit message.  Either way, I won't move until you tell me which you'd like to do.

Reply:
Feel free to write the commit message.